plugins/SlicerT/SlicerTView.cpp

```diff
@@ -178,16 +190,101 @@ void SlicerTView::dropEvent(QDropEvent* de)
+	brush.drawPixmap(QRect(0, boxTopY, s_leftBoxWidth, s_bottomBoxHeight), m_background); // left
+	brush.fillRect(
+		QRect(s_leftBoxWidth, boxTopY, width() - s_leftBoxWidth, s_bottomBoxHeight), QColor(23, 26, 31)); // right
+	brush.fillRect(QRect(0, 0, width(), s_topBarHeight), QColor(20, 23, 27));							  // top
+
+	// top bar dividers
+	brush.setPen(QColor(56, 58, 60));
+	brush.drawLine(0, s_topBarHeight - 1, width(), s_topBarHeight - 1);
+	brush.drawLine(0, 0, width(), 0);
+
+	// sample name divider
+	brush.setPen(QColor(56, 58, 60));
+	brush.drawLine(0, boxTopY, width(), boxTopY);
+
+	// boxes divider
+	brush.setPen(QColor(56, 24, 94));
+	brush.drawLine(s_leftBoxWidth, boxTopY, s_leftBoxWidth, height());
+
+	// --- top bar
+	brush.drawPixmap(
+		QRect(10, (s_topBarHeight - m_fullLogo.height()) / 2, m_fullLogo.width(), m_fullLogo.height()), m_fullLogo);
+
+	int y1_text = m_y1 + 27;
+
+	// --- left box
+	brush.setPen(QColor(255, 255, 255));
+	brush.drawText(s_x1 - 25, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Threshold"));
+	brush.drawText(s_x2 - 25, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Fade Out"));
+	brush.drawText(s_x3 - 25, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Reset"));
+	brush.drawText(s_x4 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Midi"));
+	brush.drawText(s_x5 - 16, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("BPM"));
+	brush.drawText(s_x6 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Snap"));
+
+	int kor = 15; // knob outer radius
+	int kir = 9;  // knob inner radius
+
+	// draw knob backgrounds
+	brush.setRenderHint(QPainter::Antialiasing);
+
+	// draw outer radius 2 times to make smooth
+	brush.setPen(QPen(QColor(159, 124, 223, 100), 4));
+	brush.drawArc(QRect(s_x1 - kor, m_y1, kor * 2, kor * 2), -45 * 16, 270 * 16);
+	brush.drawArc(QRect(s_x2 - kor, m_y1, kor * 2, kor * 2), -45 * 16, 270 * 16);
+
+	brush.setPen(QPen(QColor(159, 124, 223, 255), 2));
+	brush.drawArc(QRect(s_x1 - kor, m_y1, kor * 2, kor * 2), -45 * 16, 270 * 16);
+	brush.drawArc(QRect(s_x2 - kor, m_y1, kor * 2, kor * 2), -45 * 16, 270 * 16);
+
+	// inner knob circle
+	brush.setBrush(QColor(106, 90, 138));
+	brush.setPen(QColor(0, 0, 0, 0));
+	brush.drawEllipse(QPoint(s_x1, m_y1 + 15), kir, kir);
+	brush.drawEllipse(QPoint(s_x2, m_y1 + 15), kir, kir);
+
+	// current sample bar
+	brush.fillRect(QRect(0, boxTopY - s_sampleBoxHeight, width(), s_sampleBoxHeight), QColor(5, 5, 5));
+
+	brush.setPen(QColor(56, 58, 60));
+	brush.drawLine(width() - 24, boxTopY - s_sampleBoxHeight, width() - 24, boxTopY);
+
+	brush.setPen(QColor(255, 255, 255, 180));
+	brush.setFont(QFont(brush.font().family(), 8, -1, false));
+	QString sampleName = m_slicerTParent->getSampleName();
+	if (sampleName == "") { sampleName = "No sample loaded"; }
+
+	brush.drawText(5, boxTopY - s_sampleBoxHeight, width(), s_sampleBoxHeight, Qt::AlignLeft, sampleName);
```

Comment:
This seems like lots of manual layout computations. Ideally the different components would be implemented as widgets of their own which would then be put into Qt's layouts so that they can resize dynamically.

Reply:
Pretty sure that using Qt's layouts would be more work and more complex. I use that many offsets to get everything to align pixel-perfectly, so even if I used layouts, I would have to add the offset in some way or another. And the move parts are honestly the simplest, the complex part is drawing the text, dividers, gradients, knobs and backgrounds. Don't really see how that could be implemented any differently.

Reply:
I think it would definitively be possible with the right resize policies and margins. Layouts are used exactly for situations like these so that complex layout calculations do not have to be reimplemented over and over again. Using widgets in layouts might also lead to more reusable components. However, I guess it's nothing for now.